soc/nordic/nrf54h/power.c

```diff
@@ -59,30 +85,59 @@ void nrf_poweroff(void)
 	CODE_UNREACHABLE;
 }
 
-#if IS_ENABLED(CONFIG_PM_S2RAM)
-/* Resume domain after local suspend to RAM. */
-static void sys_resume(void)
+static void s2idle_enter(uint8_t substate_id)
 {
-	if (IS_ENABLED(CONFIG_ICACHE)) {
-		/* Power up and re-enable ICACHE */
-		nrf_memconf_ramblock_control_enable_set(NRF_MEMCONF, RAMBLOCK_POWER_ID,
-							RAMBLOCK_CONTROL_BIT_ICACHE, true);
-		sys_cache_instr_enable();
+	sys_snode_t *node;
+
+	node = soc_pd_sys_snode_get();
+
+	switch (substate_id) {
+	case 0: /* Substate for idle with cache retained - not implemented yet. */
```

Comment:
0 should be for "Idle" - like calling `k_cpu_idle()`
1 should be reserved for the cache retained in RAM
2 for the cache disabled

Reply:
Isn't it like, that "idle like calling `k_cpu_idle()` " is the default state performed when no circumstances for other states to be used. I noticed that when I call `k_sleep(t)`, where `t < [any of residency times defined]`, function `s2idle_enter()` is not executed, but regular idle thread is called (so [here](https://github.com/nrfconnect/sdk-zephyr/blob/main/arch/arm/core/cortex_m/cpu_idle.c#L94) So does it make sense to define such state (or substate)?

Reply:
I don't know internals of zephyr pm, if it bypasses the policy module and calls k_cpu_idle by itself in some circumstances. By still, to align the implementation with the design, I would prefer to change the substates enumeration. Even if the substate 0 would be unused

Reply:
The idle thread doesn't bypass the power module strictly speaking, but it will call `k_cpu_idle()` when `pm_suspend(ticks)` returns false, and that is if ticks is below any of the defined thresholds of the power states. See [here](https://github.com/nrfconnect/sdk-zephyr/blob/main/kernel/idle.c#L85)

Reply:
Ok, so the problem is that the "Idle" power state is not defined and not handled by the PM subsystem and because of this the idle thread falls back to `k_cpu_idle()`. So then for clear implementation, I would suggest implementing all power states in the PM so that `pm_suspend(ticks)` always returns `true`. For very short sleep, it can even select the `ACTIVE` state and return `true`.

Reply:
I don't think that'll work either, when the pm system picks the `ACTIVE` state as next state `pm_suspend()` returns `false`
https://github.com/zephyrproject-rtos/zephyr/blob/06bdd1001c7989f36731d34a5809a1ba22359d4a/subsys/pm/pm.c#L215